File: Player.js
class Player {
    constructor() {
        this.meepleCount = 7;
        this.score = 0;
    }

    placeTile(board) {
        let x = 0;
        try { x = parseInt(CarcassonneMain.inp.coordx.getText()); }
        catch (e) {
            CarcassonneMain.inp.error.setText("Invalid tile. Try new coordinates");
            CarcassonneMain.waitButton();
            this.placeTile(board);
            return;
        }
        let y = 0;
        try { y = parseInt(CarcassonneMain.inp.coordy.getText()); }
        catch (e) {
            CarcassonneMain.inp.error.setText("Invalid tile. Try new coordinates");
            CarcassonneMain.waitButton();
            this.placeTile(board);
            return;
        }
        let rotateans = 0;
        try { rotateans = parseInt(CarcassonneMain.inp.rots.getText()); }
        catch (e) {
            CarcassonneMain.inp.error.setText("Invalid rotations. Try new number");
            CarcassonneMain.waitButton();
            this.placeTile(board);
            return;
        }
        this.rotateTile(board.turnCount, rotateans, board);
        for (let j = 0; j < 4; j++) {
            // console.log(board.tiles.get(board.turnCount).types[j]);
        }
        if (x > 0 && y > 0 && (board.validSide(board.tiles.get(board.turnCount), board.board[x][y - 1], 0))
            && (board.validSide(board.tiles.get(board.turnCount), board.board[x + 1][y], 1))
            && (board.validSide(board.tiles.get(board.turnCount), board.board[x][y + 1], 2))
            && (board.validSide(board.tiles.get(board.turnCount), board.board[x - 1][y], 3))
            && (board.board[x][y].types[0] == -1)
            && ((board.board[x + 1][y].types[0] != -1) || (board.board[x][y + 1].types[0] != -1)
                || (board.board[x - 1][y].types[0] != -1) || (board.board[x][y - 1].types[0] != -1))) {

            board.board[x][y] = board.tiles.get(board.turnCount);

            let worked = false;
            let position;
            try {
                position = parseInt(CarcassonneMain.inp.meeple.getText());
            } catch (e) {
                position = -1;
            }

            if (position != -1)
                worked = this.placeMeeple(position, x, y, board);
            else
                worked = true;

            while (!worked) {
                CarcassonneMain.inp.error.setText("invalid meeple position. Choose -1 to not place one.");
                CarcassonneMain.waitButton();
                if (parseInt(CarcassonneMain.inp.meeple.getText()) != -1) {
                    try {
                        position = parseInt(CarcassonneMain.inp.meeple.getText());
                    } catch (e) {
                        position = -1;
                    }
                    if (position != -1)
                        worked = this.placeMeeple(position, x, y, board);
                    else
                        worked = true;
                } else {
                    break;
                }
            }
            if (worked) {
                this.meepleCount--;
            }
            const scorer = new Scorer(board, x, y);
            scorer.scoring(x, y, board);
        } else {
            CarcassonneMain.inp.error.setText("Invalid tile. Try new coordinates");
            CarcassonneMain.waitButton();
            this.placeTile(board);
        }
    }

    placeMeeple(position, x, y, board) {
        const tile = board.board[x][y];
        const turn = board.turnCount % board.players.length;
        tile.meeple[0] = turn;
        tile.meeple[14] = position;

        switch (position % 3) {
            case 1:
                tile.meeple[1] = board.board[x][y].types[Math.floor(position / 3)];
                if ((tile.types[(position) / 3] == 2)) {
                    tile.meeple[position + 2] = 1;
                    tile.meeple[position + 1] = 1;
                    tile.meeple[position + 3] = 1;
                    for (let j = 0; j < 3; j++) {
                        if (position / 3 <= j) {
                            if (board.board[x][y].connected[position / 3][j] && board.board[x][y].types[j + 1] == 2) {
                                tile.meeple[(j + 1) * 3 + 2 + 1] = 1;
                                tile.meeple[(j + 1) * 3 + 2] = 1;
                                tile.meeple[(j + 1) * 3 + 2 + 2] = 1;
                            }
                        } else if (board.board[x][y].connected[position / 3][j] && board.board[x][y].types[j] == 2) {
                            tile.meeple[j * 3 + 2 + 1] = 1;
                            tile.meeple[j * 3 + 2] = 1;
                            tile.meeple[j * 3 + 2 + 2] = 1;
                        }
                    }
                } else if ((tile.types[(position) / 3] == 1)) {
                    tile.meeple[position + 2] = 1;
                    for (let j = 0; j < 3; j++) {
                        if (position / 3 <= j) {
                            if (board.board[x][y].connected[position / 3][j] && board.board[x][y].types[j + 1] == 1) {
                                tile.meeple[(j + 1) * 3 + 2 + 1] = 1;
                                break;
                            }
                        } else if (board.board[x][y].connected[position / 3][j] && board.board[x][y].types[j] == 1) {
                            tile.meeple[j * 3 + 2 + 1] = 1;
                            break;
                        }
                    }
                } else {
                    board.fieldMeeples.push([x, y]);
                    tile.meeple[position + 2] = 1;
                    tile.meeple[position + 1] = 1;
                    tile.meeple[position + 3] = 1;
                    for (let j = 0; j < 3; j++) {
                        if (position / 3 <= j) {
                            if (board.board[x][y].connected[position / 3][j] && board.board[x][y].types[j + 1] == 0) {
                                tile.meeple[(j + 1) * 3 + 2 + 1] = 1;
                                tile.meeple[(j + 1) * 3 + 2] = 1;
                                tile.meeple[(j + 1) * 3 + 2 + 2] = 1;
                            }
                        } else if (board.board[x][y].connected[position / 3][j] && board.board[x][y].types[j] == 0) {
                            tile.meeple[j * 3 + 2 + 1] = 1;
                            tile.meeple[j * 3 + 2] = 1;
                            tile.meeple[j * 3 + 2 + 2] = 1;
                        }
                    }
                    if (tile.types[(position / 3 + 1) % 4] == 0) {
                        tile.meeple[(position + 2) % 12 + 2] = 1;
                    }
                    if (tile.types[(position / 3 + 3) % 4] == 0) {
                        tile.meeple[(position - 2) % 12 + 2] = 1;
                    }

                    if (tile.type == "edgecityroadleft") {
                        tile.meeple[(position + 7) % 12 + 2] = 1;
                    } else if (tile.type == "edgecityroadright") {
                        tile.meeple[(position + 5) % 12 + 2] = 1;
                    }
                }
                break;
            case 0:
                if (tile.types[(position) / 3] != 1 || tile.types[((position) / 3 - 1) % 3] == 0) {
                    tile.meeple[0] = -1;
                    tile.meeple[14] = -1;
                    return false;
                }
                board.fieldMeeples.push([x, y]);
                if (tile.types[((position) / 3 - 1) % 3] == 1) {
                    tile.meeple[position + 2] = 1;
                    tile.meeple[(position - 1) % 12 + 2] = 1;
                } else {
                    tile.meeple[position + 2] = 1;
                    if (tile.type == "cornercityroad") {
                        tile.meeple[(position + 5) % 12 + 2] = 1;
                    } else if ((tile.type == "edgecityroadt") || (tile.type == "edgecityroadstraight")) {
                        tile.meeple[(position + 8) % 12 + 2] = 1;
                    }
                }
                tile.meeple[1] = 0;
                break;
            case 2:
                if (tile.types[(position) / 3] != 1 || tile.types[((position) / 3 + 1) % 3] == 0) {
                    tile.meeple[0] = -1;
                    tile.meeple[14] = -1;
                    return false;
                }
                board.fieldMeeples.push([x, y]);
                if (tile.types[((position) / 3 + 1) % 3] == 1) {
                    tile.meeple[position + 2] = 1;
                    tile.meeple[(position + 1) % 12 + 2] = 1;
                } else {
                    tile.meeple[position + 2] = 1;
                    if (tile.type == "cornercityroad") {
                        tile.meeple[(position + 7) % 12 + 2] = 1;
                    } else if ((tile.type == "edgecityroadt") || (tile.type == "edgecityroadstraight")) {
                        tile.meeple[(position + 4) % 12 + 2] = 1;
                    }
                }
                tile.meeple[1] = 0;
                break;
            default:
                return false;
        }
        return true;
    }

    rotateTile(turn, rotate, board) {
        let old;
        let oldcon;
        let oldcono;
        board.tiles.get(turn).rotations = board.tiles.get(turn).rotations + rotate;
        for (let i = 0; i < rotate; i++) {
            old = board.tiles.get(turn).types.slice();
            board.tiles.get(turn).types = [old[3], old[0], old[1], old[2]];

            for (let j = 0; j < 4; j++) {
                oldcon = board.tiles.get(turn).connected[j];
                if (j != 3)
                    board.tiles.get(turn).connected[j] = [oldcon[2], oldcon[0], oldcon[1]];
            }

            oldcono = board.tiles.get(turn).connected;
            board.tiles.get(turn).connected = [oldcono[3], oldcono[0], oldcono[1], oldcono[2]];
        }
    }
}

File: Player.py
import CarcassonneMain, GUI, Input, Player, Scorer, Tile, EmptyTile

class Player:
    def __init__(self):
        self.meeple_count = 7
        self.score = 0

    def place_tile(self, board):
        try:
            x = int(CarcassonneMain.inp.coordx.getText())
        except (ValueError, NumberFormatException):
            CarcassonneMain.inp.error.setText("Invalid tile. Try new coordinates")
            CarcassonneMain.waitButton()
            self.place_tile(board)
            return

        try:
            y = int(CarcassonneMain.inp.coordy.getText())
        except (ValueError, NumberFormatException):
            CarcassonneMain.inp.error.setText("Invalid tile. Try new coordinates")
            CarcassonneMain.waitButton()
            self.place_tile(board)
            return

        try:
            rotate_ans = int(CarcassonneMain.inp.rots.getText())
        except (ValueError, NumberFormatException):
            CarcassonneMain.inp.error.setText("Invalid rotations. Try new number")
            CarcassonneMain.waitButton()
            self.place_tile(board)
            return

        self.rotate_tile(board.turn_count, rotate_ans, board)

        if x > 0 and y > 0 and (board.valid_side(board.tiles[board.turn_count], board.board[x][y - 1], 0)) and (board.valid_side(board.tiles[board.turn_count], board.board[x + 1][y], 1)) and (board.valid_side(board.tiles[board.turn_count], board.board[x][y + 1], 2)) and (board.valid_side(board.tiles[board.turn_count], board.board[x - 1][y], 3)) and (board.board[x][y].types[0] == -1) and ((board.board[x + 1][y].types[0] != -1) or (board.board[x][y + 1].types[0] != -1) or (board.board[x - 1][y].types[0] != -1) or (board.board[x][y - 1].types[0] != -1)):

            board.board[x][y] = board.tiles[board.turn_count]

            worked = False
            try:
                position = int(CarcassonneMain.inp.meeple.getText())
            except (ValueError, NumberFormatException):
                position = -1

            if position != -1:
                worked = self.place_meeple(position, x, y, board)
            else:
                worked = True

            while not worked:
                CarcassonneMain.inp.error.setText("invalid meeple position. Choose -1 to not place one.")
                CarcassonneMain.waitButton()
                if int(CarcassonneMain.inp.meeple.getText()) != -1:
                    try:
                        position = int(CarcassonneMain.inp.meeple.getText())
                    except (ValueError, NumberFormatException):
                        position = -1
                    if position != -1:
                        worked = self.place_meeple(position, x, y, board)
                    else:
                        worked = True
                else:
                    break

            if worked:
                self.meeple_count -= 1

            scorer = Scorer(board, x, y)
            scorer.scoring(x, y, board)

        else:
            CarcassonneMain.inp.error.setText("Invalid tile. Try new coordinates")
            CarcassonneMain.waitButton()
            self.place_tile(board)

    def place_meeple(self, position, x, y, board):
        tile = board.board[x][y]
        turn = board.turn_count % len(board.players)
        tile.meeple[0] = turn
        tile.meeple[14] = position

        switch = {
            1: self.handle_edge_case,
            0: self.handle_side_case,
            2: self.handle_side_case
        }
        case_handler = switch.get(position % 3, lambda: False)
        return case_handler(position, x, y, board, tile)

    def handle_edge_case(self, position, x, y, board, tile):
        tile.meeple[1] = board.board[x][y].types[position // 3]
        if tile.types[position // 3] == 2:
            tile.meeple[position + 2] = 1
            tile.meeple[position + 1] = 1
            tile.meeple[position + 3] = 1
            for j in range(3):
                if position // 3 <= j:
                    if board.board[x][y].connected[position // 3][j] and board.board[x][y].types[j + 1] == 2:
                        tile.meeple[(j + 1) * 3 + 2 + 1] = 1
                        tile.meeple[(j + 1) * 3 + 2] = 1
                        tile.meeple[(j + 1) * 3 + 2 + 2] = 1
                elif board.board[x][y].connected[position // 3][j] and board.board[x][y].types[j] == 2:
                    tile.meeple[j * 3 + 2 + 1] = 1
                    tile.meeple[j * 3 + 2] = 1
                    tile.meeple[j * 3 + 2 + 2] = 1
        elif tile.types[position // 3] == 1:
            tile.meeple[position + 2] = 1
            for j in range(3):
                if position // 3 <= j:
                    if board.board[x][y].connected[position // 3][j] and board.board[x][y].types[j + 1] == 1:
                        tile.meeple[(j + 1) * 3 + 2 + 1] = 1
                        break
                elif board.board[x][y].connected[position // 3][j] and board.board[x][y].types[j] == 1:
                    tile.meeple[j * 3 + 2 + 1] = 1
                    break
        else:
            board.field_meeples.append([x, y])
            tile.meeple[position + 2] = 1
            tile.meeple[position + 1] = 1
            tile.meeple[position + 3] = 1
            for j in range(3):
                if position // 3 <= j:
                    if board.board[x][y].connected[position // 3][j] and board.board[x][y].types[j + 1] == 0:
                        tile.meeple[(j + 1) * 3 + 2 + 1] = 1
                        tile.meeple[(j + 1) * 3 + 2] = 1
                        tile.meeple[(j + 1) * 3 + 2 + 2] = 1
                elif board.board[x][y].connected[position // 3][j] and board.board[x][y].types[j] == 0:
                    tile.meeple[j * 3 + 2 + 1] = 1
                    tile.meeple[j * 3 + 2] = 1
                    tile.meeple[j * 3 + 2 + 2] = 1
            if tile.types[(position // 3 + 1) % 4] == 0:
                tile.meeple[(position + 2) % 12 + 2] = 1
            if tile.types[(position // 3 + 3) % 4] == 0:
                tile.meeple[(position - 2) % 12 + 2] = 1

            if tile.type == "edgecityroadleft":
                tile.meeple[(position + 7) % 12 + 2] = 1
            elif tile.type == "edgecityroadright":
                tile.meeple[(position + 5) % 12 + 2] = 1
        return True

    def handle_side_case(self, position, x, y, board, tile):
        if tile.types[position // 3] != 1 or tile.types[(position // 3 - 1) % 3] == 0:
            tile.meeple[0] = -1
            tile.meeple[14] = -1
            return False
        board.field_meeples.append([x, y])
        if tile.types[(position // 3 - 1) % 3] == 1:
            tile.meeple[position + 2] = 1
            tile.meeple[(position - 1) % 12 + 2] = 1
        else:
            tile.meeple[position + 2] = 1
            if tile.type == "cornercityroad":
                tile.meeple[(position + 5) % 12 + 2] = 1
            elif tile.type == "edgecityroadt" or tile.type == "edgecityroadstraight":
                tile.meeple[(position + 8) % 12 + 2] = 1
        tile.meeple[1] = 0
        return True

    def rotate_tile(self, turn, rotate, board):
        board.tiles[turn].rotations += rotate
        for _ in range(rotate):
            old = board.tiles[turn].types[:]
            board.tiles[turn].types = [old[3], old[0], old[1], old[2]]

            for j in range(4):
                oldcon = board.tiles[turn].connected[j][:]
                if j != 3:
                    board.tiles[turn].connected[j] = [oldcon[2], oldcon[0], oldcon[1]]

            oldcono = board.tiles[turn].connected[:]
            board.tiles[turn].connected = [oldcono[3], oldcono[0], oldcono[1], oldcono[2]]

